state-sync/src/coordinator.rs

```diff
@@ -811,92 +869,97 @@ impl<T: ExecutorProxyTrait> StateSyncCoordinator<T> {
-            .ok_or_else(|| format_err!("Expected version has overflown!"))?;
+            .ok_or_else(|| Error::IntegerOverflow("Expected version has overflown!".into()))?;
         if chunk_start_version != expected_version {
-            // Old / wrong chunk.
             self.request_manager.process_chunk_version_mismatch(
                 peer,
                 chunk_start_version,
                 known_version,
             )?;
         }
 
-        let chunk_size = txn_list_with_proof.len() as u64;
+        // Process the chunk based on the response type
         match response.response_li {
             ResponseLedgerInfo::VerifiableLedgerInfo(li) => {
-                self.process_response_with_verifiable_li(txn_list_with_proof, li, None)
+                self.process_response_with_verifiable_li(txn_list_with_proof.clone(), li, None)
             }
             ResponseLedgerInfo::ProgressiveLedgerInfo {
                 target_li,
                 highest_li,
             } => {
                 let highest_li = highest_li.unwrap_or_else(|| target_li.clone());
-                ensure!(
-                    target_li.ledger_info().version() <= highest_li.ledger_info().version(),
-                    "Progressive ledger info received target LI {} higher than highest LI {}",
-                    target_li,
-                    highest_li
-                );
-                self.process_response_with_verifiable_li(
-                    txn_list_with_proof,
-                    target_li,
-                    Some(highest_li),
-                )
+                if target_li.ledger_info().version() > highest_li.ledger_info().version() {
+                    let error_message = format!(
+                        "Progressive ledger info received a target LI {} higher than highest LI {}",
+                        target_li, highest_li
+                    );
+                    Err(Error::ProcessInvalidChunk(error_message).into())
+                } else {
+                    self.process_response_with_verifiable_li(
+                        txn_list_with_proof.clone(),
+                        target_li,
+                        Some(highest_li),
+                    )
+                }
             }
             ResponseLedgerInfo::LedgerInfoForWaypoint {
                 waypoint_li,
                 end_of_epoch_li,
             } => self.process_response_with_waypoint_li(
-                txn_list_with_proof,
+                txn_list_with_proof.clone(),
                 waypoint_li,
                 end_of_epoch_li,
             ),
         }
-        .map_err(|e| {
+        .map_err(|error| {
             self.request_manager.process_invalid_chunk(&peer);
-            format_err!("[state sync] failed to apply chunk: {}", e)
+            Error::ProcessInvalidChunk(format!("{}", error))
         })?;
 
+        // Update counters and logs with processed chunk information
+        let chunk_size = txn_list_with_proof.len() as u64;
```

Comment:
Here, we could have gotten the chunk size in it's original place, rather than clone the `txn_list_with_proof` as nothing changes between those points

Reply:
See my response above.